### DISCUSSION TITLE
Form the output filename dynamically based on macro definitions in the tex file

### DISCUSSION THREAD

**Discussion by urmel1960** [WinEdt-Team/WinEdt-Support#14] [1.1 General Help and Discussion]
Hi there

I am using WinEDT 11 (latest build) and I would like to dynamically form the output filename of the PDF to be generated (using lualatex) based on macro definitions from the main LaTeX file.

Example:

In the file main.tex, the following macros are defined in the preamble:

```
\newcommand{\foo01}{xxxxx}
\newcommand{\foo02}{yyyyy}
\newcommand{\foo03}{zzzzz}
```

By compiling main.tex, the output file should be named as follows:

`xxxxx-yyyyy-zzzzz.pdf`

At the moment I have no clue how to do this. Any ideas and advice is highly appreciated.

Regards, Albrecht

**Comment by Alex-Simonic** [comment]
This is very complicated and possibly problematic. For starter, you can put the following in your preamble:

```
% !TEX program = lualatex
% !PDFTeXify engine = lualatex
% !lualatex ext = --jobname="xxxx-yyyy-zzzz"
% !PDFTeXify ext = --jobname="xxxx-yyyy-zzzz"
```
This will compile to xxxxx-yyyyy-zzzzz.pdf with lualatex or PDFTeXify (using lualatex).

But now you have opened a can of worms as WinEdt preview macros are not aware of the name change. If this does what you want it is possible to specify a permanent switch for lualatex in Execution Modes with

`%["Exe('%b\Exec\Jobname.edt');"]%-9`

and write a macro Jobname.edt that parses the preamble with tagged RegEx (pretty much what WinEdt compile macros do with pragma comments) and assigns register -9 to empty or --jobname="foo" depending on what it finds in your preamble. Then it would take a lot of testing to iron out any problems and it would almost certainly need further modifications to other (preview) macros to make it work as you want or expect.

This is too complicated and problematic for defaults and I will not work on it. The above hint may or may not allow you to proceed or get help from somebody else who is familiar with WinEdt macros and RegEx and has some time to spare.

**Comment by urmel1960** [comment]
@Alex-Simonic  Thank you very much for your input. It has given me some important food for thought. 

The background to my query is as follows:

We are a consulting company and create a wide variety of documents for our clients as part of our mandates. We use LaTex for various reasons, such as 
- high and reliably reproducible output quality
- simple management of the versions in progress via a self-hosted GitLab instance
- .....

The documents are of course versioned and have corresponding metadata as is usual in a certified QM management system.

It would therefore be nice if the file name of the generated PDF automatically reflects the respective version,  status, date and also some other metadata. Up to now, we have renamed the output files manually, which is annoying and can also be a source of errors. 

Your justified concerns about possible problems with my original approach have prompted me to look for another solution. 

I am now thinking about copying the PDF after compilation and renaming it automatically using a powershell script. The script should read the required metadata from the preamble of the tex file and then rename the resulting PDF accordingly. 

For this, I would like to be able to start a powershell script automatically after compiling. 

Can I do this in any way by configuring WinEDT accordingly ?

**Reply by Alex-Simonic** [reply]
Yes, this can be done. But your question is still a bit vague as it is far from clear how the naming convention should be implemented. This is what I would do:

1. Create a local macro `%b\Exec\CopyPDF.edt`. For testing purpose, you can just put in a line:
`
Prompt("Rename PDF: %P%N.pdf");
`
2. Options Interface, Advanced Configuration, Compile Macros: Compiler Engine. Depending on your compile options (starting the viewer, etc) you will have to find the spot where your custom processing should happen. For example, before starting the pdf viewer:

```
:PDF_Preview:: ======================================================
  IfStr("%$('%!9-View_Search');","1","=",>
   !|Exe('%b\Exec\PDF\PDF Search.edt');|,>
   !|Exe('%b\Exec\PDF\PDF Preview.edt');|);
  JMP(!'Exit');
```
Here you can add:
`
  Exe("%b\Exec\CopyPDF.edt");
`
This macro can then Run PowerShell to do the job or it can use WinEdt macros to determine the output filename and Copy the file (whatever you are more comfortable with). The details depend on whether you want to preview the copy or the original or neither...

**Reply by urmel1960** [reply]
That is exactly what I was looking for.  Thank you very much. 

I would leave the original way of work as it is - that means I simply would create an additional,  renamed copy (according to my naming conventions) and would leave the preview of the pdf with the original as it is out of the box .
